import os.path
import sys
import getopt
import csv
import operator
from collections import defaultdict
import inspect



# Global variables
IDBA_UD_HOME=""

# Main 
def main(argv):
    blastn()


def blastn():
    print_status("calling blastn()")
    
def blastp():
    print_status("calling blastp()")
    
def blastx():
    print_status("calling blastx()")    

def filter_blast():
    print_status("filtering blast results")    
  

def prepare_reference_genome():
    print_status("Preparing reference genomes")    
    


def run_ESOM():
    print_status("Preparing reference genomes")    
      


# Preprocess FASTQ
def preprocess():
    print_status("calling preprocess()")  


# IDBA-UD
def run_idba_ud():
    print_status("run IDBA-UD")

# 
def parse_opt():
    print_status("calling parsing opts")


# Print the usage of this script 
def print_usage():
    print("An integrated pipeline for processing metagenomic sequencing reads.")
    print(" ")
    print("Usage:")
    print("  python pick_contig.py -i BLAST-RESULT-INFILE -o FILTER-OUTFILE [-q] [-b BITSCORE-CUTOFF] [-l ALIGNMENT-LENGTH-CUTOFF] [-p PERCENTAGE-IDENTITY]")
    print("      -i STRING  Input file generated by BLAST with -m6 option")
    print(" ")
    print(" ") 
    print("Ver 0.2")


# 
def import_settings():
    print_status("importing settings")


# Print status
def print_status(msg):
    caller_name = inspect.stack()[1][3]
    msg = "[ " + caller_name + " ]  " + msg
    print msg
    

# Invoke the main function
if __name__ == "__main__":
    main(sys.argv[1:])